definition_validator/validator.cc

```diff
@@ -189,6 +189,42 @@ void validateOverriding(const core::GlobalState &gs, core::SymbolRef method) {
     }
 }
 
+void validateFinal(const core::GlobalState &gs, core::SymbolRef klass, optional<core::SymbolRef> nonSingleton) {
+    auto parent = klass.data(gs)->superClass();
+    if (!nonSingleton.has_value() && parent.exists() && parent.data(gs)->isClassFinal()) {
+        if (auto e = gs.beginError(klass.data(gs)->loc(), core::errors::Resolver::FinalAncestor)) {
+            e.setHeader("`{}` was declared as final and cannot be inherited by `{}`", parent.data(gs)->show(gs),
+                        klass.data(gs)->show(gs));
+        }
+    }
```

Comment:
Why does this check need to know whether `nonSingleton` exists? It seems like even if it's the singleton class, the parent class would be `Class`, which would not be final, and so this error wouldn't be emitted.

Reply:
Removing the check for existence of `nonSingleton` results in this:

```
$ cat foo.rb 
# typed: true
class Final
  extend T::Helpers
  final!
end

class Bad < Final; end
$ sorbet foo.rb 
foo.rb:7: Final was declared as final and cannot be inherited by Bad https://srb.help/5047
     7 |class Bad < Final; end
        ^^^^^^^^^^^^^^^^^

foo.rb:7: T.class_of(Final) was declared as final and cannot be inherited by T.class_of(Bad) https://srb.help/5047
     7 |class Bad < Final; end
              ^^^
Errors: 2
```